README
NAME
    HTTP::Exception - throw HTTP-Errors as (Exception::Class-) Exceptions

VERSION
    0.03001

INSTALLATION
    To install this module, run the following commands:

        perl Build.PL
        ./Build
        ./Build test
        ./Build install

SYNOPSIS
    HTTP::Exception lets you throw HTTP-Errors as Exceptions.

        use HTTP::Exception;

        # throw a 404 Exception
        HTTP::Exception->throw(404);

        # later in your framework
        eval { ... };
        if (my $e = HTTP::Exception->caught) {
            # do some errorhandling stuff
            print $e->code;             # 404
            print $e->status_message;   # Not Found
        }

    You can also throw HTTP::Exception-subclasses like this.

        # same 404 Exception
        eval { HTTP::Exception::404->throw(); };
        eval { HTTP::Exception::NOT_FOUND->throw(); };

    And catch them accordingly.

        # same 404 Exception
        eval { HTTP::Exception::404->throw(); };

        if (my $e = HTTP::Exception::405->caught)       { do stuff } # won't catch
        if (my $e = HTTP::Exception::404->caught)       { do stuff } # will catch
        if (my $e = HTTP::Exception::NOT_FOUND->caught) { do stuff } # will catch
        if (my $e = HTTP::Exception::4XX->caught)       { do stuff } # will catch all 4XX Exceptions
        if (my $e = HTTP::Exception->caught)            { do stuff } # will catch every HTTP::Exception
        if (my $e = Exception::Class->caught)           { do stuff } # catch'em all

    You can create Exceptions and not throw them, because maybe you want to
    set some fields manually. See HTTP::Exception and HTTP::Exception for
    more info.

        # is not thrown, ie doesn't die, only created
        my $e = HTTP::Exception->new(404);

        # usual stuff works
        $e->code;               # 404
        $e->status_message      # Not Found

        # set status_message to something else
        $e->status_message('Nothing Here')

        # fails, because code is only an accessor, see section ACCESSORS below
        # $e->code(403);

        # and finally throw our prepared exception
        $e->throw;

DESCRIPTION
    Every HTTP::Exception is a Exception::Class - Class. So the same
    mechanisms apply as with Exception::Class-classes. In fact have a look
    at Exception::Class' docs for more general information on exceptions and
    Exception::Class::Base for information on what methods a caught
    exception also has.

    HTTP::Exception is only a factory for HTTP::Exception::XXX (where X is a
    number) subclasses. That means that HTTP::Exception->new(404) returns a
    HTTP::Exception::404 object, which in turn is a HTTP::Exception::Base -
    Object.

    Don't bother checking a caught HTTP::Exception::...-class with "isa" as
    it might not contain what you would expect. Use the code- or
    status_message-attributes and the is_ -methods instead.

    The subclasses are created at compile-time, ie the first time you make
    "use HTTP::Exception". See paragraph below for the naming scheme of
    those subclasses.

    Subclassing the subclasses works as expected.

NAMING SCHEME
  HTTP::Exception::XXX
    X is a Number and XXX is a valid HTTP-Statuscode. All HTTP-Statuscodes
    are supported. See chapter HTTP::Exception

  HTTP::Exception::STATUS_MESSAGE
    STATUS_MESSAGE is the same name as a HTTP::Status Constant WITHOUT the
    HTTP_ at the beginning. So see HTTP::Status for more details.

IMPORTING SPECIFIC ERROR RANGES
    It is possible to load only specific ranges of errors. For example

        use HTTP::Exception qw(5XX);

        HTTP::Exception::500->throw; # works
        HTTP::Exception::400->throw; # won't work anymore

    will only create HTTP::Exception::500 till HTTP::Exception::510. In
    theory this should save some memory, but I don't have any numbers, that
    back up this claim.

    You can load multiple ranges

        use HTTP::Exception qw(3XX 4XX 5XX);

    And there are aliases for ranges

        use HTTP::Exception qw(CLIENT_ERROR)

    The following aliases exist and load the specified ranges:

        REDIRECTION   => 3XX
        CLIENT_ERROR  => 4XX
        SERVER_ERROR  => 5XX
        ERROR         => 4XX 5XX
        ALL           => 1XX 2XX 3XX 4XX 5XX

    And of course, you can load multiple aliased ranges

        use HTTP::Exception qw(REDIRECTION ERROR)

    ALL is the same as not specifying any specific range.

        # the same
        use HTTP::Exception qw(ALL);
        use HTTP::Exception;

ACCESSORS (READONLY)
  code
    A valid HTTP-Statuscode. See HTTP::Status for information on what codes
    exist.

  is_info
    Return TRUE if `$self-'code> is an *Informational* status code (1xx).
    This class of status code indicates a provisional response which can't
    have any content.

  is_success
    Return TRUE if `$self-'code> is a *Successful* status code (2xx).

  is_redirect
    Return TRUE if `$self-'code> is a *Redirection* status code (3xx). This
    class if status code indicates that further action needs to be taken by
    the user agent in order to fulfill the request.

  is_error
    Return TRUE if `$self-'code> is an *Error* status code (4xx or 5xx). The
    function return TRUE for both client error or a server error status
    codes.

  is_client_error
    Return TRUE if `$self-'code> is an *Client Error* status code (4xx).
    This class of status code is intended for cases in which the client
    seems to have erred.

  is_server_error
    Return TRUE if `$self-'code> is an *Server Error* status code (5xx).
    This class of status codes is intended for cases in which the server is
    aware that it has erred or is incapable of performing the request.

    *POD for is_ methods is Copy/Pasted from HTTP::Status, so check back
    there and alert me of changes.*

FIELDS
    Fields are the same as ACCESSORS except they can be set. Either you set
    them during Exception creation (->new) or Exception throwing (->throw).

        HTTP::Exception->new(200, status_message => "Everything's fine");
        HTTP::Exception::200->new(status_message => "Everything's fine");
        HTTP::Exception::OK->new(status_message => "Everything's fine");

        HTTP::Exception->throw(200, status_message => "Everything's fine");
        HTTP::Exception::200->throw(status_message => "Everything's fine");
        HTTP::Exception::OK->throw(status_message => "Everything's fine");

    Catch them in your Webframework like this

        eval { ... }
        if (my $e = HTTP::Exception->caught) {
            print $e->code;          # 200
            print $e->status_message # "Everything's fine" instead of the usual ok
        }

  status_message
    DEFAULT The HTTP-Statusmessage as provided by HTTP::Status

    A Message, that represents the Execptions' Status for Humans.

PLACK
    HTTP::Exception can be used with Plack::Middleware::HTTPExceptions. But
    HTTP::Exception does not depend on Plack, you can use it anywhere else.
    It just plays nicely with Plack.

COMPLETENESS
    For the sake of completeness, HTTP::Exception provides exceptions for
    non-error-http-statuscodes. This means you can do

        HTTP::Exception->throw(200);

    which throws an Exception of type OK. Maybe useless, but complete. A
    more realworld-example would be a redirection

        # all are exactly the same
        HTTP::Exception->throw(301, location => 'google.com');
        HTTP::Exception::301->throw(location => 'google.com');
        HTTP::Exception::MOVED_PERMANENTLY->throw(location => 'google.com');

CAVEATS
    The HTTP::Exception-Subclass-Creation relies on HTTP::Status. It's
    possible that the Subclasses change, when HTTP::Status' constants are
    changed.

    New Subclasses are created automatically, when constants are added to
    HTTP::Status. That means in turn, that Subclasses disappear, when
    constants are removed from HTTP::Status.

    The Changes-File of HTTP::Status indicates, that the last change to its
    constants was made in 2008. I think, that breaking changes are quite
    unlikely.

AUTHOR
    Thomas Mueller, `<tmueller at cpan.org>'

SEE ALSO
  Exception::Class, Exception::Class::Base
    Consult Exception::Class' documentation for the Exception-Mechanism and
    Exception::Class::Base' docs for a list of methods our caught Exception
    is also capable of.

  HTTP::Status
    Constants, Statuscodes and Statusmessages

  Plack, especially Plack::Middleware::HTTPExceptions
    Have a look at Plack, because it rules in general. In the first place,
    this Module was written as the companion for
    Plack::Middleware::HTTPExceptions, but since it doesn't depend on Plack,
    you can use it anywhere else, too.

BUGS
    Please report any bugs or feature requests to `bug-http-exception at
    rt.cpan.org', or through the web interface at
    http://rt.cpan.org/NoAuth/ReportBug.html?Queue=HTTP-Exception. I will be
    notified, and then you'll automatically be notified of progress on your
    bug as I make changes.

SUPPORT
    You can find documentation for this module with the perldoc command.

        perldoc HTTP::Exception

    You can also look for information at:

    * RT: CPAN's request tracker
        http://rt.cpan.org/NoAuth/Bugs.html?Dist=HTTP-Exception

    * AnnoCPAN: Annotated CPAN documentation
        http://annocpan.org/dist/HTTP-Exception

    * CPAN Ratings
        http://cpanratings.perl.org/d/HTTP-Exception

    * Search CPAN
        http://search.cpan.org/dist/HTTP-Exception/

LICENSE AND COPYRIGHT
    Copyright 2010 Thomas Mueller.

    This program is free software; you can redistribute it and/or modify it
    under the terms of either: the GNU General Public License as published
    by the Free Software Foundation; or the Artistic License.

    See http://dev.perl.org/licenses/ for more information.



Changes
{{$NEXT}}

0.04007   2018-07-18 22:49:22+02:00 Europe/Berlin
    GitHub PR #2 merged (thanks karenetheridge)
        * link to HTTP::Throwable    
    GitHub PR #4 merged (thanks timbunce)
        * use q{} for _status_message so message can contain a single quote
        * throw exception on loader code error instead of silently ignoring
    GitHub PR #5 merged (thanks neilb)
        * Added [MetaJSON] so releases will include META.json
        * Added "use warnings" to all modules

0.04006   2014-02-14 00:27:50+01:00 Europe/Berlin
    typo fix (thanks dsteinbrunner)

0.04005   2014-02-14 00:03:43+01:00 Europe/Berlin
    corrected Encoding in dist.ini,
        due to CPAN Testers reporting an invalid byte sequence in the Makefile

0.04004   2012-08-17 01:00:11 Europe/Berlin
    Dist::Zilla::Plugin::GitHub::Meta extracts correct Metainfos now

0.04003   2012-08-17 00:38:56 Europe/Berlin
    Versions are now handled by Dist::Zilla::Plugin::PkgVersion
    deleted 2 unused testclasses
    
0.04002   2012-08-17 00:08:19 Europe/Berlin
    RT 79021: Eliminate warnings about "subroutine redefined"
        * subroutine redefined warnings under mod_perl silenced
        * created testcase "emulating" mod_perl
    removed shebang in authortests

0.04001   2012-02-24 22:55:01 Europe/Berlin
    RT 75271: AutoPrereq of Dist::Zilla found too many dependencies,
        * fixed by stating dependencies explicitly
        (thanks jquelin)

0.04      2012-02-18 22:04:20 Europe/Berlin
    HTTP::Status 6.03 added a status, that broke this module
    moved to Dist::Zilla

0.03001 2010-03-11
    Fixed tests not to emit warnings anymore

0.03000 2010-03-11
    Added possibility to import only specific error ranges
    Added tests and Pod for this

0.02006 2010-03-08
    Added minimum version requirements to prerequisites

0.02005 2010-03-05
    Added Pod VERSION sections back in, thanks to ShipIt::Step::ChangePodVersion ;) 

0.02004 2010-03-04
    Corrected $VERSION for changed modules

0.02003 2010-03-04
    Removed parent.pm - dependency
    Added base.pm - dependency, works better and is longer in Core
    HTTP::Request::Common required for t/05-plack.t, othwerwise skipped

0.02002 2010-03-04
    Some small POD enhancements for H::E::Base and H::E::Loader

0.02001 2010-03-04
    Fixed POD for HTTP::Exception::nXX-Classes (again copy/paste)
    Updated README

0.02000 2010-03-03
    Added "Tween-Classes" to the hierarchy (1XX .. 5XX)
    Added Tests for "Tween-Classes" to the hierarchy (1XX .. 5XX)
    Fixed t/06-is_methods.t not to use Plack::Test anymore
    Changed HTTP::Exception::Base to use as_string from Exception::Class::Base
    Added location field for 3XX Exceptions
    Added message and error as synonym for status_message (compat for Exception::Class)
    Fixed t/06-is_methods.t to not include Plack::Test anymore (evil copy/paste)
    Forgot dates in the Changes-File :\

0.01007 2010-03-03
    fixed some POD

0.01006 2010-03-02
    changed EMail-Address in packages to my CPAN-EMail

0.01005 2010-03-02
    removed VERSION from pod, too much hassle to keep in sync with $VERSION

0.01004 2010-03-02
    still trying to shipit

0.01003 2010-03-02
    no changes to the module itself, just trying to get shipit to work

0.01002 2010-03-02
    removed MANIFEST.SKIP from MANIFEST
    Corrected BuildRequirements in Build.PL (added Test::Exception)
    Corrected Requirements in Makefile.PL, thought it is recreated automatically
    fixed t/05.plack to be skipped on systems without Plack

0.01001 2010-03-02
    Corrected Requirements in Build.PL

0.01000 2010-02-29
    First Implementation of HTTP::Exceptions


lib/HTTP/Exception.pm
package HTTP::Exception;

use strict;
use warnings;
use HTTP::Status;
use Scalar::Util qw(blessed);

################################################################################
sub import {
    my ($class) = shift;
    require HTTP::Exception::Loader;
    HTTP::Exception::Loader->import(@_);
}

# act as a kind of factory here
sub new  {
    my $class       = shift;
    my $error_code  = shift;

    die ('HTTP::Exception->throw needs a HTTP-Statuscode to throw')  unless ($error_code);
    die ("Unknown HTTP-Statuscode: $error_code") unless (HTTP::Status::status_message ($error_code));

    "HTTP::Exception::$error_code"->new(@_);
}

# makes HTTP::Exception->caught possible instead of HTTP::Exception::Base->caught
sub caught {
    my $self = shift;
    my $e = $@;
    return $e if (blessed $e && $e->isa('HTTP::Exception::Base'));
    $self->SUPER::caught(@_);
}

1;


=head1 NAME

HTTP::Exception - throw HTTP-Errors as (Exception::Class-) Exceptions

=begin readme

=head1 INSTALLATION

To install this module, run the following commands:

    perl Build.PL
    ./Build
    ./Build test
    ./Build install

=end readme

=head1 SYNOPSIS

HTTP::Exception lets you throw HTTP-Errors as Exceptions.

    use HTTP::Exception;

    # throw a 404 Exception
    HTTP::Exception->throw(404);

    # later in your framework
    eval { ... };
    if (my $e = HTTP::Exception->caught) {
        # do some errorhandling stuff
        print $e->code;             # 404
        print $e->status_message;   # Not Found
    }

You can also throw HTTP::Exception-subclasses like this.

    # same 404 Exception
    eval { HTTP::Exception::404->throw(); };
    eval { HTTP::Exception::NOT_FOUND->throw(); };

And catch them accordingly.

    # same 404 Exception
    eval { HTTP::Exception::404->throw(); };

    if (my $e = HTTP::Exception::405->caught)       { do stuff } # won't catch
    if (my $e = HTTP::Exception::404->caught)       { do stuff } # will catch
    if (my $e = HTTP::Exception::NOT_FOUND->caught) { do stuff } # will catch
    if (my $e = HTTP::Exception::4XX->caught)       { do stuff } # will catch all 4XX Exceptions
    if (my $e = HTTP::Exception->caught)            { do stuff } # will catch every HTTP::Exception
    if (my $e = Exception::Class->caught)           { do stuff } # catch'em all

You can create Exceptions and not throw them, because maybe you want to set some
fields manually. See L<HTTP::Exception/"FIELDS"> and
L<HTTP::Exception/"ACCESSORS"> for more info.

    # is not thrown, ie doesn't die, only created
    my $e = HTTP::Exception->new(404);

    # usual stuff works
    $e->code;               # 404
    $e->status_message      # Not Found

    # set status_message to something else
    $e->status_message('Nothing Here')

    # fails, because code is only an accessor, see section ACCESSORS below
    # $e->code(403);

    # and finally throw our prepared exception
    $e->throw;

=head1 DESCRIPTION

Every HTTP::Exception is a L<Exception::Class> - Class. So the same mechanisms
apply as with L<Exception::Class>-classes. In fact have a look at
L<Exception::Class>' docs for more general information on exceptions and
L<Exception::Class::Base> for information on what methods a caught exception
also has.

HTTP::Exception is only a factory for HTTP::Exception::XXX (where X is a number)
subclasses. That means that HTTP::Exception->new(404) returns a
HTTP::Exception::404 object, which in turn is a HTTP::Exception::Base - Object.

Don't bother checking a caught HTTP::Exception::...-class with "isa" as it might
not contain what you would expect. Use the code- or status_message-attributes
and the is_ -methods instead.

The subclasses are created at compile-time, ie the first time you make
"use HTTP::Exception". See paragraph below for the naming scheme of those
subclasses.

Subclassing the subclasses works as expected.

=head1 NAMING SCHEME

=head2 HTTP::Exception::XXX

X is a Number and XXX is a valid HTTP-Statuscode. All HTTP-Statuscodes are
supported. See chapter L<HTTP::Exception/"COMPLETENESS">

=head2 HTTP::Exception::STATUS_MESSAGE

STATUS_MESSAGE is the same name as a L<HTTP::Status> Constant B<WITHOUT>
the HTTP_ at the beginning. So see L<HTTP::Status/"CONSTANTS"> for more details.

=head1 IMPORTING SPECIFIC ERROR RANGES

It is possible to load only specific ranges of errors. For example

    use HTTP::Exception qw(5XX);

    HTTP::Exception::500->throw; # works
    HTTP::Exception::400->throw; # won't work anymore

will only create HTTP::Exception::500 till HTTP::Exception::510. In theory this
should save some memory, but I don't have any numbers, that back up this claim.

You can load multiple ranges

    use HTTP::Exception qw(3XX 4XX 5XX);

And there are aliases for ranges

    use HTTP::Exception qw(CLIENT_ERROR)

The following aliases exist and load the specified ranges:

    REDIRECTION   => 3XX
    CLIENT_ERROR  => 4XX
    SERVER_ERROR  => 5XX
    ERROR         => 4XX 5XX
    ALL           => 1XX 2XX 3XX 4XX 5XX

And of course, you can load multiple aliased ranges

    use HTTP::Exception qw(REDIRECTION ERROR)

ALL is the same as not specifying any specific range.

    # the same
    use HTTP::Exception qw(ALL);
    use HTTP::Exception;

=head1 ACCESSORS (READONLY)

=head2 code

A valid HTTP-Statuscode. See L<HTTP::Status> for information on what codes exist.

=head2 is_info

Return TRUE if C<$self->code> is an I<Informational> status code (1xx).  This
class of status code indicates a provisional response which can't have
any content.

=head2 is_success

Return TRUE if C<$self->code> is a I<Successful> status code (2xx).

=head2 is_redirect

Return TRUE if C<$self->code> is a I<Redirection> status code (3xx). This class
if status code indicates that further action needs to be taken by the
user agent in order to fulfill the request.

=head2 is_error

Return TRUE if C<$self->code> is an I<Error> status code (4xx or 5xx).  The
function return TRUE for both client error or a server error status codes.

=head2 is_client_error

Return TRUE if C<$self->code> is an I<Client Error> status code (4xx). This
class of status code is intended for cases in which the client seems to
have erred.

=head2 is_server_error

Return TRUE if C<$self->code> is an I<Server Error> status code (5xx). This
class of status codes is intended for cases in which the server is aware
that it has erred or is incapable of performing the request.

I<POD for is_ methods is Copy/Pasted from L<HTTP::Status>, so check back there and
alert me of changes.>

=head1 FIELDS

Fields are the same as ACCESSORS except they can be set. Either you set them
during Exception creation (->new) or Exception throwing (->throw).

    HTTP::Exception->new(200, status_message => "Everything's fine");
    HTTP::Exception::200->new(status_message => "Everything's fine");
    HTTP::Exception::OK->new(status_message => "Everything's fine");

    HTTP::Exception->throw(200, status_message => "Everything's fine");
    HTTP::Exception::200->throw(status_message => "Everything's fine");
    HTTP::Exception::OK->throw(status_message => "Everything's fine");

Catch them in your Webframework like this

    eval { ... }
    if (my $e = HTTP::Exception->caught) {
        print $e->code;          # 200
        print $e->status_message # "Everything's fine" instead of the usual ok
    }

=head2 status_message

B<DEFAULT> The HTTP-Statusmessage as provided by L<HTTP::Status>

A Message, that represents the Execptions' Status for Humans.

=head1 PLACK

HTTP::Exception can be used with L<Plack::Middleware::HTTPExceptions>. But
HTTP::Exception does not depend on L<Plack>, you can use it anywhere else. It
just plays nicely with L<Plack>.

=head1 COMPLETENESS

For the sake of completeness, HTTP::Exception provides exceptions for
non-error-http-statuscodes. This means you can do

    HTTP::Exception->throw(200);

which throws an Exception of type OK. Maybe useless, but complete.
A more realworld-example would be a redirection

    # all are exactly the same
    HTTP::Exception->throw(301, location => 'google.com');
    HTTP::Exception::301->throw(location => 'google.com');
    HTTP::Exception::MOVED_PERMANENTLY->throw(location => 'google.com');

=head1 CAVEATS

The HTTP::Exception-Subclass-Creation relies on L<HTTP::Status>.
It's possible that the Subclasses change, when HTTP::Status'
constants are changed.

New Subclasses are created automatically, when constants are added to
HTTP::Status. That means in turn, that Subclasses disappear, when constants
are removed from L<HTTP::Status>.

Some constants were added to L<HTTP::Status>' in February 2012. As a result
HTTP::Exception broke. But that was the result of uncareful coding on my side.
I think, that breaking changes are now quite unlikely.

=head1 AUTHOR

Thomas Mueller, C<< <tmueller at cpan.org> >>

=head1 SEE ALSO

=head2 L<Exception::Class>, L<Exception::Class::Base>

Consult Exception::Class' documentation for the Exception-Mechanism and
Exception::Class::Base' docs for a list of methods our caught Exception is also
capable of.

=head2 L<HTTP::Status>

Constants, Statuscodes and Statusmessages

=head2 L<HTTP::Throwable>, built on top of the more modern L<Throwable> framework (the successor to L<Exception::Class>)

=head2 L<Plack>, especially L<Plack::Middleware::HTTPExceptions>

Have a look at Plack, because it rules in general. In the first place, this
Module was written as the companion for L<Plack::Middleware::HTTPExceptions>,
but since it doesn't depend on Plack, you can use it anywhere else, too.

=head1 BUGS

Please report any bugs or feature requests to
C<bug-http-exception at rt.cpan.org>, or through the web interface at
L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=HTTP-Exception>.
I will be notified, and then you'll automatically be notified of progress on
your bug as I make changes.

=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc HTTP::Exception

You can also look for information at:

=over 4

=item * RT: CPAN's request tracker

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=HTTP-Exception>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/HTTP-Exception>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/HTTP-Exception>

=item * Search CPAN

L<https://metacpan.org/release/HTTP-Exception>

=back

=head1 LICENSE AND COPYRIGHT

Copyright 2010 Thomas Mueller.

This program is free software; you can redistribute it and/or modify it
under the terms of either: the GNU General Public License as published
by the Free Software Foundation; or the Artistic License.

See http://dev.perl.org/licenses/ for more information.


=cut
